server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java

```diff
@@ -343,10 +345,42 @@ default void deQueueShouldWaitForAMailToBeEnqueued(ExecutorService executorServi
         assertThat(tryDequeue.get().getMail().getName()).isEqualTo("name");
     }
 
+    @Test
+    default void concurrentEnqueueDequeueShouldNotFail() throws Exception {
+        MailQueue testee = getMailQueue();
+
+        ConcurrentLinkedDeque<Mail> dequeuedMails = new ConcurrentLinkedDeque<>();
+
+        int threadCount = 10;
+        int operationCount = 100;
+        int totalDequeuedMessages = 500;
+        ConcurrentTestRunner.builder()
+            .operation((threadNumber, step) -> {
+                if (step % 2 == 0) {
+                    testee.enQueue(defaultMail()
+                        .name("name" + threadNumber + "-" + step)
+                        .build());
+                } else {
+                    MailQueue.MailQueueItem mailQueueItem = testee.deQueue();
+                    dequeuedMails.add(mailQueueItem.getMail());
+                    mailQueueItem.done(true);
+                }
+            })
+            .threadCount(threadCount)
+            .operationCount(operationCount)
+            .runSuccessfullyWithin(1, TimeUnit.MINUTES);
+
+        assertThat(
+            dequeuedMails.stream()
+                .map(Mail::getName)
+                .distinct())
+            .hasSize(totalDequeuedMessages);
```

Comment:
I would like to have another assertion mailqueue now is empty

Reply:
I can not as this is part of the regular mailQueue. I can not access the management capabilities to easily do this change.

We could assert that dequeue timeout but this is a rather complicated assertion, whose intent won't be clear.